src/likelihoods/negativebinomial.jl

```diff
@@ -1,36 +1,109 @@
 
-```math
-    p(k|successes, f) = \\frac{\\Gamma(k+successes)}{k! \\Gamma(successes)} l(f)^successes (1 - l(f))^k
+To create a new parametrization, you need to;
+- Create a new type `struct MyNBParam{T} <: NBParam; myparams::T; end`
+- Dispatch `(l::NegativeBinomialLikelihood{<:MyNBParam})(f::Real)`, which return a [`NegativeBinomial`](https://juliastats.org/Distributions.jl/latest/univariate/#Distributions.NegativeBinomial) from `Distributions.jl`.
+`NegativeBinomial` follows the parametrization of [`NBParamI`](@ref), i.e. the first argument is the number of successes
+and the second argument is the probability of success.
+
+## Examples
+
+```julia-repl
+julia> lik = NegativeBinomialLikelihood(NBParamI(10), logistic)
+NegativeBinomialLikelihood{NBParamI{Int64}, LogisticLink}(NBParamI{Int64}(10), LogisticLink(LogExpFunctions.logistic))
+julia> lik(2.0)
+Distributions.NegativeBinomial{Float64}(r=10.0, p=0.8807970779778824)
+
+julia> lik = NegativeBinomialLikelihood(NBParamIII(10), exp)
+NegativeBinomialLikelihood{NBParamIII{Int64}, ExpLink}(NBParamIII{Int64}(10), ExpLink(exp))
+julia> lik(2.0)
+Distributions.NegativeBinomial{Float64}(r=10.0, p=0.4249256576603398)
 ```
-On calling, this returns a negative binomial distribution with `successes` successes and 
-probability of success equal to `l(f)`.
-
-!!! warning "Parameterization" 
-    The parameter `successes` is different from the parameter `r` in the 
-    [Wikipedia definition](http://en.wikipedia.org/wiki/Negative_binomial_distribution), 
-    which denotes the number of failures.
-    This parametrization is used in order to stay consistent with the parametrization in 
-    [Distributions.jl](https://juliastats.org/Distributions.jl/stable/univariate/#Distributions.NegativeBinomial).
-    To use the Wikipedia definition, set `successes` as the number of "failures" and
-    change the probability of success from `l(f)` to `1 - l(f)`.
-    Note that with symmetric functions like the [`LogisticLink`](@ref), this corresponds to
-    using `l(-f)`.
-"""
-struct NegativeBinomialLikelihood{Tl<:AbstractLink,T<:Real} <: AbstractLikelihood
-    successes::T    # number of successes parameter
+"""
+struct NegativeBinomialLikelihood{Tp<:NBParam,Tl<:AbstractLink} <: AbstractLikelihood
+    params::Tp # Likelihood parametrization (and parameters)
     invlink::Tl
+    function NegativeBinomialLikelihood(params::Tparam, invlink) where {Tparam}
+        invlink = link(invlink)
+        return new{Tparam,typeof(invlink)}(params, invlink)
+    end
 end
 
-function NegativeBinomialLikelihood(l=logistic; successes::Real=1)
-    return NegativeBinomialLikelihood(successes, link(l))
+@deprecate NegativeBinomialLikelihood(l=logistic; successes=1) NegativeBinomialLikelihood(
+    NBParamI(successes), l
+)
+
+function (l::NegativeBinomialLikelihood)(::Real)
+    return error(
+        "not implemented for type $(typeof(l)). See `NegativeBinomialLikelihood` docs"
```

Comment:
How is this different from just not implementing a method for some type ? how would it be possible to trigger this?

Reply:
So if someone writes a custom `NBParam`, and try to run it he would see this error. I agree it could be a bit more verbose.